castle/cms/passwordvalidation/hibp.py

```diff
@@ -0,0 +1,49 @@
+from castle.cms.interfaces.passwordvalidation import ICustomPasswordValidator
+from plone import api
+from zope.interface import implementer
+import hashlib
+import urllib2
+
+from castle.cms import _
+
+
+@implementer(ICustomPasswordValidator)
+class HaveIBeenPwned(object):
+    """ Checks if the password has already been used
+    by iterating over the memberdata property password_history_size.
+    """
+
+    def __init__(self, context):
+        self.context = context
+
+    def validate(self, password, data):
+        # Permit empty passwords here to allow registration links.
+        # Plone will force the user to set one.
+        if password is None:
+            return None
+
+        if api.user.is_anonymous():
+            # No check for registrations.
+            return None
+
+        hibp_enabled = api.portal.get_registry_record(
+            'plone.hibp_enabled'
+        )
+
+        if not hibp_enabled:
+            # hibp has been disabled.
+            return None
+
+        sha = hashlib.sha1(password).hexdigest()
+        needle = sha[5:].upper()
+        url = 'https://api.pwnedpasswords.com/range/' + sha[:5]
```

Comment:
We're already using the requests library if you want to use that instead of urllib2. This is fine as well though.

Reply:
Oh I had no idea that was already being used. Yeah I'd rather use that.